Ideas for speeding up skia-bindings

I seems like the build script from `skia-bindings` is slowing down my dev tools. 

Each time I run `cargo check`, it takes some time to complete. 

It's been some time, but I've written a number of C bindings for rust before, such as [ffmpeg-dev-rs](https://github.com/imager-io/ffmpeg-dev-rs) and [x264-dev](https://github.com/imager-io/x264-dev). If I can recall correctly, what I resorted to then was: if you're in debug mode and if the generated dependencies already exist, then I'll short circuit compilation in the `build.rs file`, to speed things up. 

Would you accept a poll request that does something similar?

Oh thanks for the writeup! 

Also I was wondering why`skia-bindings/src/bindings.rs` didn't just reference the generated code VIA `include!(concat!(env!("OUT_DIR"), "/bindings.rs"));`. 

Regarding 
> What do you exactly mean by that, the bindings.rs file and the bindings library? But what happens then if the feature set or some dependent environment variables change?

and
> Yes, of course, if this works reliably, but how exactly does the short circuit mechanism work?

Honestly, I just started experimenting with something rather naive [over at my fork](https://github.com/colbyn/rust-skia). 

By naive, I mean: 
```rust
fn main() {
    // Attempt to speed up compilation while in debug mode. . 
    if env::is_debug_mode() {
        let lib_file_exists = build_support::cargo::output_directory()
            .join("skia/libskia.a")
            .exists();
        let bindings_rs_file_exisst = std::env::current_dir()
            .map(|pwd| pwd.join("src/bindings.rs"))
            .map_or(false, |path| path.exists());
        log(format!(
            "lib_file_exists={} bindings_rs_file_exisst={} output_directory={:?}\n",
            lib_file_exists,
            bindings_rs_file_exisst,
            build_support::cargo::output_directory()
        ));
        if lib_file_exists && bindings_rs_file_exisst {
            return;
        }
    }
    …
}
```

Regarding "To avoid that, I make sure to run VSCode from the shell", this is something I always do. But perhaps it's different if vscode is restarted. 

I still don't know what's causing the build script to be reevaluated... This is what happens when I run `cargo check` twice on rust-skia/master: 
```shell
➜  rust-skia git:(master) ✗ cargo check
    Blocking waiting for file lock on build directory
   Compiling skia-bindings v0.41.0 (/Users/colbyn/Scratch/rust-skia/skia-bindings)
warning: In file included from src/bindings.cpp:19:
warning: In file included from /Users/colbyn/Scratch/rust-skia/skia-bindings/skia/include/core/SkDeferredDisplayListRecorder.h:11:
warning: In file included from /Users/colbyn/Scratch/rust-skia/skia-bindings/skia/include/core/SkDeferredDisplayList.h:12:
warning: /Users/colbyn/Scratch/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:211:49: warning: unused parameter 'width' [-Wunused-parameter]
warning:     SkSurfaceCharacterization createResized(int width, int height) const {
warning:                                                 ^
warning: /Users/colbyn/Scratch/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:211:60: warning: unused parameter 'height' [-Wunused-parameter]
warning:     SkSurfaceCharacterization createResized(int width, int height) const {
warning:                                                            ^
warning: /Users/colbyn/Scratch/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:223:47: warning: unused parameter 'usesGLFBO0' [-Wunused-parameter]
warning:     SkSurfaceCharacterization createFBO0(bool usesGLFBO0) const {
warning:                                               ^
warning: /Users/colbyn/Scratch/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:227:54: warning: unused parameter 'other' [-Wunused-parameter]
warning:     bool operator==(const SkSurfaceCharacterization& other) const { return false; }
warning:                                                      ^
warning: 4 warnings generated.
    Checking skia-safe v0.41.0 (/Users/colbyn/Scratch/rust-skia/skia-safe)
    Checking skia-org v1.0.0 (/Users/colbyn/Scratch/rust-skia/skia-org)
    Finished dev [unoptimized + debuginfo] target(s) in 57.22s
➜  rust-skia git:(master) ✗ cargo check
   Compiling skia-bindings v0.41.0 (/Users/colbyn/Scratch/rust-skia/skia-bindings)
warning: In file included from src/bindings.cpp:19:
warning: In file included from /Users/colbyn/Scratch/rust-skia/skia-bindings/skia/include/core/SkDeferredDisplayListRecorder.h:11:
warning: In file included from /Users/colbyn/Scratch/rust-skia/skia-bindings/skia/include/core/SkDeferredDisplayList.h:12:
warning: /Users/colbyn/Scratch/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:211:49: warning: unused parameter 'width' [-Wunused-parameter]
warning:     SkSurfaceCharacterization createResized(int width, int height) const {
warning:                                                 ^
warning: /Users/colbyn/Scratch/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:211:60: warning: unused parameter 'height' [-Wunused-parameter]
warning:     SkSurfaceCharacterization createResized(int width, int height) const {
warning:                                                            ^
warning: /Users/colbyn/Scratch/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:223:47: warning: unused parameter 'usesGLFBO0' [-Wunused-parameter]
warning:     SkSurfaceCharacterization createFBO0(bool usesGLFBO0) const {
warning:                                               ^
warning: /Users/colbyn/Scratch/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:227:54: warning: unused parameter 'other' [-Wunused-parameter]
warning:     bool operator==(const SkSurfaceCharacterization& other) const { return false; }
warning:                                                      ^
warning: 4 warnings generated.
    Checking skia-safe v0.41.0 (/Users/colbyn/Scratch/rust-skia/skia-safe)
    Checking skia-org v1.0.0 (/Users/colbyn/Scratch/rust-skia/skia-org)
    Finished dev [unoptimized + debuginfo] target(s) in 27.77s
➜  rust-skia git:(master) ✗
```

# Environment 
- OS: Macos Bug Sur (version `11.4`)
- The currently active `rustc` version is `rustc 1.52.1

I can reproduce this on my Mac, you are right, `cargo check` runs every time, taking ~14 seconds:
``` zsh
rust-skia % cargo check
   Compiling skia-bindings v0.41.0 (/Users/armin/dev/rust-skia/skia-bindings)
warning: In file included from src/bindings.cpp:19:
warning: In file included from /Users/armin/dev/rust-skia/skia-bindings/skia/include/core/SkDeferredDisplayListRecorder.h:11:
warning: In file included from /Users/armin/dev/rust-skia/skia-bindings/skia/include/core/SkDeferredDisplayList.h:12:
warning: /Users/armin/dev/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:211:49: warning: unused parameter 'width' [-Wunused-parameter]
warning:     SkSurfaceCharacterization createResized(int width, int height) const {
warning:                                                 ^
warning: /Users/armin/dev/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:211:60: warning: unused parameter 'height' [-Wunused-parameter]
warning:     SkSurfaceCharacterization createResized(int width, int height) const {
warning:                                                            ^
warning: /Users/armin/dev/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:223:47: warning: unused parameter 'usesGLFBO0' [-Wunused-parameter]
warning:     SkSurfaceCharacterization createFBO0(bool usesGLFBO0) const {
warning:                                               ^
warning: /Users/armin/dev/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:227:54: warning: unused parameter 'other' [-Wunused-parameter]
warning:     bool operator==(const SkSurfaceCharacterization& other) const { return false; }
warning:                                                      ^
warning: 4 warnings generated.
    Checking skia-safe v0.41.0 (/Users/armin/dev/rust-skia/skia-safe)
    Checking skia-org v1.0.0 (/Users/armin/dev/rust-skia/skia-org)
    Finished dev [unoptimized + debuginfo] target(s) in 13.98s
rust-skia % cargo check
   Compiling skia-bindings v0.41.0 (/Users/armin/dev/rust-skia/skia-bindings)
warning: In file included from src/bindings.cpp:19:
warning: In file included from /Users/armin/dev/rust-skia/skia-bindings/skia/include/core/SkDeferredDisplayListRecorder.h:11:
warning: In file included from /Users/armin/dev/rust-skia/skia-bindings/skia/include/core/SkDeferredDisplayList.h:12:
warning: /Users/armin/dev/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:211:49: warning: unused parameter 'width' [-Wunused-parameter]
warning:     SkSurfaceCharacterization createResized(int width, int height) const {
warning:                                                 ^
warning: /Users/armin/dev/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:211:60: warning: unused parameter 'height' [-Wunused-parameter]
warning:     SkSurfaceCharacterization createResized(int width, int height) const {
warning:                                                            ^
warning: /Users/armin/dev/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:223:47: warning: unused parameter 'usesGLFBO0' [-Wunused-parameter]
warning:     SkSurfaceCharacterization createFBO0(bool usesGLFBO0) const {
warning:                                               ^
warning: /Users/armin/dev/rust-skia/skia-bindings/skia/include/core/SkSurfaceCharacterization.h:227:54: warning: unused parameter 'other' [-Wunused-parameter]
warning:     bool operator==(const SkSurfaceCharacterization& other) const { return false; }
warning:                                                      ^
warning: 4 warnings generated.
    Checking skia-safe v0.41.0 (/Users/armin/dev/rust-skia/skia-safe)
    Checking skia-org v1.0.0 (/Users/armin/dev/rust-skia/skia-org)
    Finished dev [unoptimized + debuginfo] target(s) in 14.56s
```

--- a/skia-bindings/build_support/cargo.rs
+++ b/skia-bindings/build_support/cargo.rs
@@ -14,21 +14,21 @@ pub fn output_directory() -> PathBuf {
     PathBuf::from(env::var("OUT_DIR").unwrap())
 }
 
-pub fn rerun_if_changed(path: impl AsRef<Path>) {
+pub fn rerun_if_file_changed(path: impl AsRef<Path>) {
     println!("cargo:rerun-if-changed={}", path.as_ref().to_str().unwrap());
 }
 
 /// Returns the value of an environment variable and notify cargo that the build
-/// should rereun if it changes.
+/// should re-run if it changes.
 pub fn env_var(name: impl AsRef<str>) -> Option<String> {
     let name = name.as_ref();
-    rerun_if_env_changed(name);
+    rerun_if_env_var_changed(name);
     env::var(name).ok()
 }
 
 /// Notify cargo that it should rerun the build if the environment
 /// variable changes.
-pub fn rerun_if_env_changed(name: impl AsRef<str>) {
+pub fn rerun_if_env_var_changed(name: impl AsRef<str>) {
     println!("cargo:rerun-if-env-changed={}", name.as_ref())
 }
 

--- a/skia-bindings/build_support/skia/config.rs
+++ b/skia-bindings/build_support/skia/config.rs
@@ -226,7 +226,7 @@ impl FinalBuildConfiguration {
                             set_target = false;
                             // Add macOS specific environment variables that affect the output of a
                             // build.
-                            cargo::rerun_if_changed("MACOSX_DEPLOYMENT_TARGET");
+                            cargo::rerun_if_env_var_changed("MACOSX_DEPLOYMENT_TARGET");
                             "mac"
                         }
                         "windows" => "win",

--- a/skia-bindings/build_support/skia_bindgen.rs
+++ b/skia-bindings/build_support/skia_bindgen.rs
@@ -161,12 +161,12 @@ pub fn generate_bindings(build: &FinalBuildConfiguration, output_directory: &Pat
     for source in &build.binding_sources {
         cc_build.file(source);
         let source = source.to_str().unwrap();
-        cargo::rerun_if_changed(source);
+        cargo::rerun_if_file_changed(source);
         builder = builder.header(source);
     }
 
     let include_path = &build.skia_source_dir;
-    cargo::rerun_if_changed(include_path.join("include"));
+    cargo::rerun_if_file_changed(include_path.join("include"));
 
     builder = builder.clang_arg(format!("-I{}", include_path.display()));
     cc_build.include(include_path);

